sshd-core/src/main/java/org/apache/sshd/common/config/keys/OpenSSHKeyUtils.java

```diff
@@ -0,0 +1,159 @@
+        return getFingerPrint(getDefaultFingerPrintFactory(), key);
+    }
+
+    /**
+     * @param f   The {@link Factory} to create the {@link Digest} to use
+     * @param key the public key - ignored if {@code null}
+     * @return the fingerprint or {@code null} if no key.
+     * <B>Note:</B> if exception encountered then returns the exception's simple class name
+     * @see #getFingerPrint(Digest, PublicKey)
+     */
+    public static String getFingerPrint(Factory<? extends Digest> f, PublicKey key) throws Exception {
+        return getFingerPrint(f.create(), key);
+    }
+
+    /**
+     * @param d   The {@link Digest} to use
+     * @param key the public key - ignored if {@code null}
+     * @return the fingerprint or {@code null} if no key.
+     * @throws Exception if cannot create fingerprint.
+     */
+    public static String getFingerPrint(Digest d, PublicKey key) throws Exception {
+        if (key == null) {
+            return null;
+        }
+
+        Buffer buffer = new ByteArrayBuffer();
+        buffer.putRawPublicKey(key);
+
+        d.init();
+        d.update(buffer.array(), 0, buffer.wpos());
+
+        byte[] data = d.digest();
+
+        String algo = d.getAlgorithm();
+        if (BuiltinDigests.md5.getAlgorithm().equals(algo)) {
+            return algo + ":" + BufferUtils.printHex(':', data).toLowerCase();
+        } else {
+            return algo.replace("-", "").toUpperCase() + ":" + Base64.encodeToString(data).replaceAll("=", "");
+        }
+    }
+
+    /**
+     * @param expected The expected fingerprint if {@code null} or empty then returns a failure with the
+     * default fingerprint.
+     * @param key the public key - if {@code null} then returns null.
+     * @return Pair<Boolean, String> - first is success indicator, second is actual fingerprint,
+     * {@code null} if no key.
+     */
+    public static Pair<Boolean, String> checkFingerPrint(String expected, PublicKey key) throws Exception {
+        if (key == null) {
+            return null;
+        }
+
+        if (GenericUtils.isEmpty(expected)) {
+            return new Pair<>(false, getFingerPrint(key));
+        }
+
+        String comps[] = expected.split(":", 2);
+        if (GenericUtils.length(comps) < 2) {
+            return new Pair<>(false, getFingerPrint(key));
+        }
+
+        DigestFactory factory;
+        // We know that all digests have a length > 2 - if 2 (or less) then assume a pure HEX value
+        if (comps[0].length() > 2) {
+            factory = BuiltinDigests.fromString(comps[0]);
+            if (factory == null) {
+                return new Pair<>(false, getFingerPrint(key));
+            }
+        } else {
+            factory = BuiltinDigests.md5;
+            expected = factory.getName() + expected.toLowerCase();
+        }
+
+        String fingerprint = getFingerPrint(factory, key);
+        return new Pair<>(expected.equals(fingerprint), fingerprint);
+    }
```

Comment:
There is still a bug if the input already contains "MD5:0A:B3:D5:F3...." - then your _expected.equals(...)_ will fail. Therefore the code should be:
```java
if (comps[0].length() > 2) {
   ...same as before...
} else {  // we know that all digest names are more than 2 letters so we assume pure HEX MD5 here
    factory = BuiltinDigests.md5;
    expected = factory.getName().toUpperCase() + ":" + expected; 
}

String fingerprint = getFingerPrint(factory, key);
boolean matches = BuiltinDigests.md5.getFactoryName().equals(factory.getName()) ? expected.equalsIgnoreCase(fingerprint) : expected.equals(fingerprint);
return return new Pair<>(matches, fingerpint);
```
Note that you also forgot to add the ':' separator and make the factory uppercase in case of implied MD5... In this context, your code is willing to accept "sha1:..." as well as "SHA1:...." since _BuiltinDigests.fromString(...)_ is case *insensitive* - is it OK or not ? Need to decide